Microsoft.Toolkit.Uwp.UI/Cache/CacheBase.cs

```diff
@@ -0,0 +1,380 @@
+                {
+                    _concurrentTasks.Add(fileName, request);
+                }
+            }
+
+            try
+            {
+                t = await request.Task;
+            }
+            catch (Exception ex)
+            {
+                System.Diagnostics.Debug.WriteLine(ex.Message);
+                if (throwOnError)
+                {
+                    throw ex;
+                }
+            }
+            finally
+            {
+                lock (_concurrencyLock)
+                {
+                    if (_concurrentTasks.ContainsKey(fileName))
+                    {
+                        _concurrentTasks.Remove(fileName);
+                    }
+                }
+            }
+
+            return t;
+        }
+
+        private async Task<T> GetFromCacheOrDownloadAsync(Uri uri, string fileName, bool preCacheOnly)
+        {
+            StorageFile baseFile = null;
+            T t = default(T);
+            DateTime expirationDate = DateTime.Now.Subtract(CacheDuration);
+
+            if (_inMemoryFileStorage.MaxItemCount > 0)
+            {
+                var msi = _inMemoryFileStorage.GetItem(fileName, CacheDuration);
+                if (msi != null)
+                {
+                    t = msi.Item;
+                }
+            }
+
+            if (t != null)
+            {
+                return t;
+            }
+
+            var folder = await GetCacheFolderAsync();
+
+            baseFile = await folder.TryGetItemAsync(fileName) as StorageFile;
+            if (await IsFileOutOfDate(baseFile, expirationDate))
+            {
+                baseFile = await folder.CreateFileAsync(fileName, CreationCollisionOption.ReplaceExisting);
+                try
+                {
+                    t = await DownloadFileAsync(uri, baseFile, preCacheOnly);
+                }
+                catch (Exception)
+                {
+                    await baseFile.DeleteAsync().AsTask().ConfigureAwait(false);
+                    throw; // rethrowing the exception changes the stack trace. just throw
+                }
+            }
+
+            if (EqualityComparer<T>.Default.Equals(t, default(T)) && !preCacheOnly)
+            {
+                using (var fileStream = await baseFile.OpenAsync(FileAccessMode.Read))
+                {
+                    t = await InitializeTypeAsync(fileStream);
+                }
+
+                if (_inMemoryFileStorage.MaxItemCount > 0)
+                {
+                    var properties = await baseFile.GetBasicPropertiesAsync().AsTask().ConfigureAwait(false);
+
+                    var msi = new InMemoryStorageItem<T>(fileName, properties.DateModified.DateTime, t);
+                    _inMemoryFileStorage.SetItem(msi);
+                }
+            }
+
+            return t;
+        }
+
+        private async Task<T> DownloadFileAsync(Uri uri, StorageFile baseFile, bool preCacheOnly)
+        {
+            T t = default(T);
```

Comment:
T type I have no problem with, it's the t variable I think should have a little more. Anyone else with any thoughts? @deltakosh?